src/hydro_system.hpp

```diff
@@ -583,6 +573,95 @@ void HydroSystem<problem_t>::FlattenShocks(amrex::MultiFab const &q_mf, amrex::M
+		amrex::Real const v_abs = std::sqrt(vx1 * vx1 + vx2 * vx2 + vx3 * vx3);
+		amrex::Real Etot = state[bx](i, j, k, energy_index);
+		amrex::Real Eint = state[bx](i, j, k, internalEnergy_index);
+		amrex::Real Ekin = rho * vsq / 2.;
+		amrex::Real rho_new = rho;
+
+		if (rho < rho_floor) {
+			rho_new = rho_floor;
+			state[bx](i, j, k, density_index) = rho_new;
+			state[bx](i, j, k, internalEnergy_index) = Eint * rho_new / rho;
+			state[bx](i, j, k, energy_index) = rho_new * vsq / 2. + (Etot - Ekin);
+			if (nscalars_ > 0) {
+				for (int n = 0; n < nscalars_; ++n) {
+					state[bx](i, j, k, scalar0_index + n) *= rho / rho_new;
+				}
+			}
+		}
+
+		if (v_abs > speedCeiling) {
+			amrex::Real rescale_factor = speedCeiling / v_abs;
+			state[bx](i, j, k, x1Momentum_index) *= rescale_factor;
+			state[bx](i, j, k, x2Momentum_index) *= rescale_factor;
+			state[bx](i, j, k, x3Momentum_index) *= rescale_factor;
+		}
+
+		// Enforcing Limits on temperature estimated from Etot and Ekin
+		// re-obtain Ekin and Etot for putting limits on Temperature
+		Ekin = std::pow(state[bx](i, j, k, x1Momentum_index), 2.) / state[bx](i, j, k, density_index) / 2.;
+		Ekin += std::pow(state[bx](i, j, k, x2Momentum_index), 2.) / state[bx](i, j, k, density_index) / 2.;
+		Ekin += std::pow(state[bx](i, j, k, x3Momentum_index), 2.) / state[bx](i, j, k, density_index) / 2.;
+		Etot = state[bx](i, j, k, energy_index);
+		amrex::Real primTemp = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, (Etot - Ekin));
+
+		if (primTemp > tempCeiling) {
+			amrex::Real prim_eint = quokka::EOS<problem_t>::ComputeEintFromTgas(state[bx](i, j, k, density_index), tempCeiling);
+			state[bx](i, j, k, energy_index) = Ekin + prim_eint;
+		}
+
+		if (primTemp < tempFloor) {
+			amrex::Real prim_eint = quokka::EOS<problem_t>::ComputeEintFromTgas(state[bx](i, j, k, density_index), tempFloor);
+			state[bx](i, j, k, energy_index) = Ekin + prim_eint;
+		}
+
+		// Enforcing Limits on Auxiliary temperature estimated from Eint
+		Eint = state[bx](i, j, k, internalEnergy_index);
+		amrex::Real auxTemp = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, Eint);
+
+		if (auxTemp > tempCeiling) {
+			state[bx](i, j, k, internalEnergy_index) = quokka::EOS<problem_t>::ComputeEintFromTgas(state[bx](i, j, k, density_index), tempCeiling);
+			state[bx](i, j, k, energy_index) = Ekin + state[bx](i, j, k, internalEnergy_index);
+		}
+
+		if (auxTemp < tempFloor) {
+			state[bx](i, j, k, internalEnergy_index) = quokka::EOS<problem_t>::ComputeEintFromTgas(state[bx](i, j, k, density_index), tempFloor);
+			state[bx](i, j, k, energy_index) = Ekin + state[bx](i, j, k, internalEnergy_index);
+		}
+
+		if (!HydroSystem<problem_t>::is_eos_isothermal()) {
+			// recompute gas energy (to prevent P < 0)
+			amrex::Real const Eint_star = Etot - 0.5 * rho_new * vsq;
+			amrex::Real const P_star = Eint_star * (HydroSystem<problem_t>::gamma_ - 1.);
+			amrex::Real P_new = P_star;
+			if (P_star < P_floor) {
```

Comment:
If we have a temperature floor, then I don't think we need a pressure floor. @markkrumholz thoughts?